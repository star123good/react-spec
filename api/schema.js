// @flow
/**
 * The combined schema out of types and resolvers (queries, mutations and subscriptions)
 */
const {
  makeExecutableSchema,
  addSchemaLevelResolveFunction,
} = require('graphql-tools');
const debug = require('debug')('api:resolvers');
const logExecutions = require('graphql-log')({
  logger: debug,
});
const { merge } = require('lodash');
import UserError from './utils/UserError';

const scalars = require('./types/scalars');
const generalTypes = require('./types/general');

const Thread = require('./types/Thread');
const Channel = require('./types/Channel');
const ChannelSlackSettings = require('./types/ChannelSlackSettings');
const Community = require('./types/Community');
const CommunitySlackSettings = require('./types/CommunitySlackSettings');
const Message = require('./types/Message');
const Reaction = require('./types/Reaction');
const User = require('./types/User');
const DirectMessageThread = require('./types/DirectMessageThread');
const Notification = require('./types/Notification');
const Meta = require('./types/Meta');
const Invoice = require('./types/Invoice');
const CommunityMember = require('./types/CommunityMember');
const ThreadParticipant = require('./types/ThreadParticipant');

const ThreadQueries = require('./queries/thread');
const channelQueries = require('./queries/channel');
const communityQueries = require('./queries/community');
const messageQueries = require('./queries/message');
const userQueries = require('./queries/user');
const reactionQueries = require('./queries/reaction');
const directMessageThreadQueries = require('./queries/directMessageThread');
const notificationQueries = require('./queries/notification');
const metaQueries = require('./queries/meta');
const communityMemberQueries = require('./queries/communityMember');
const communitySlackSettingsQueries = require('./queries/communitySlackSettings');
const channelSlackSettingsQueries = require('./queries/channelSlackSettings');

const messageMutations = require('./mutations/message');
const threadMutations = require('./mutations/thread');
const reactionMutations = require('./mutations/reaction');
const communityMutations = require('./mutations/community');
const channelMutations = require('./mutations/channel');
const notificationMutations = require('./mutations/notification');
const userMutations = require('./mutations/user');
const metaMutations = require('./mutations/meta');
const fileMutations = require('./mutations/files');

const rateLimit = require('./utils/rate-limit-directive').default;

const IS_PROD = process.env.NODE_ENV === 'production' && !process.env.FORCE_DEV;

const Root = /* GraphQL */ `
  directive @rateLimit(
    max: Int
    window: Int
    message: String
    identityArgs: [String]
    arrayLengthField: String
  ) on FIELD_DEFINITION

  # The dummy queries and mutations are necessary because
  # graphql-js cannot have empty root types and we only extend
  # these types later on
  # Ref: apollographql/graphql-tools#293
  type Query {
    dummy: String
  }

  type Mutation {
    dummy: String
  }

  type Subscription {
    dummy: String
  }

  schema {
    query: Query
    mutation: Mutation
    subscription: Subscription
  }
`;

const resolvers = merge(
  {},
  //queries
  scalars.resolvers,
  ThreadQueries,
  channelQueries,
  communityQueries,
  messageQueries,
  userQueries,
  directMessageThreadQueries,
  reactionQueries,
  notificationQueries,
  metaQueries,
  communityMemberQueries,
  communitySlackSettingsQueries,
  channelSlackSettingsQueries,
  // mutations
  messageMutations,
  threadMutations,
  reactionMutations,
  communityMutations,
  channelMutations,
  notificationMutations,
  userMutations,
  metaMutations,
  fileMutations
);

if (process.env.NODE_ENV === 'development' && debug.enabled) {
  logExecutions(resolvers);
}

// Create the final GraphQL schema out of the type definitions
// and the resolvers
const schema = makeExecutableSchema({
  typeDefs: [
    scalars.typeDefs,
    generalTypes,
    Root,
    Community,
    CommunitySlackSettings,
    CommunityMember,
    Channel,
    ChannelSlackSettings,
    Thread,
    ThreadParticipant,
    Message,
    Reaction,
    User,
    DirectMessageThread,
    Notification,
    Meta,
    Invoice,
  ],
  resolvers,
  schemaDirectives: IS_PROD
    ? {
        rateLimit,
      }
    : {},
});

if (process.env.REACT_APP_MAINTENANCE_MODE === 'enabled') {
  console.error('\n\n⚠️ ----MAINTENANCE MODE ENABLED----⚠️\n\n');
  addSchemaLevelResolveFunction(schema, () => {
    throw new UserError(
      "We're currently undergoing planned maintenance. We'll be back soon, please check https://twitter.com/withspectrum for ongoing updates!"
    );
  });
}

module.exports = schema;
