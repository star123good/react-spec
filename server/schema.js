const { makeExecutableSchema } = require('graphql-tools');

const Story = require('./types/Story');
const queries = require('./queries');

const Query = /* GraphQL */ `
	# Root Query
	type Query {
		stories: [Story!]
	}
`;

const Schema = /* GraphQL */ `
	schema {
		query: Query
	}
`;

module.exports = makeExecutableSchema({
  typeDefs: [Schema, Query, Story],
  resolvers: {
    Query: queries,
  },
});
