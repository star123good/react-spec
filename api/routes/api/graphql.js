// @flow
import { graphqlExpress } from 'graphql-server-express';
import depthLimit from 'graphql-depth-limit';
import costAnalysis from 'graphql-cost-analysis';
import UserError from '../../utils/UserError';
import createLoaders from '../../loaders/';
import createErrorFormatter from '../../utils/create-graphql-error-formatter';
import schema from '../../schema';
import { track } from 'shared/analytics';

export default graphqlExpress(req => {
  const loaders = createLoaders();

  let currentUser = req.user && !req.user.bannedAt ? req.user : null;

  /*
    Simplify tracking calls in mutations. If a mutation causes multiple
    tracking events for different users, a userId field in the props
    will become the user id tagged in downstream analytics providers
  */
  const trackContext = (event: string, props: Object = {}) => {
    const { userId, ...rest } = props;
    const user = userId || req.user.id;
    return track(user, event, rest);
  };

  return {
    schema,
    formatError: createErrorFormatter(req),
    context: {
      loaders,
      user: currentUser,
      track: trackContext,
    },
    validationRules: [
      depthLimit(10),
      costAnalysis({
        variables: req.body.variables,
        maximumCost: 750,
        defaultCost: 1,
        createError: (max, actual) => {
          const err = new UserError(
            `GraphQL query exceeds maximum complexity, please remove some nesting or fields and try again. (max: ${max}, actual: ${actual})`
          );
          return err;
        },
      }),
    ],
  };
});
