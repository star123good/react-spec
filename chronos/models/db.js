/**
 * Database setup is done here
 */
const fs = require('fs');
const path = require('path');
const IS_PROD = !process.env.FORCE_DEV && process.env.NODE_ENV === 'production';

const DEFAULT_CONFIG = {
  db: 'spectrum',
};

const PRODUCTION_CONFIG = {
  password: process.env.COMPOSE_RETHINKDB_PASSWORD,
  host: process.env.COMPOSE_RETHINKDB_URL,
  port: process.env.COMPOSE_RETHINKDB_PORT,
  ssl: {
    ca: IS_PROD && require('raw-loader!../../cacert'),
  },
};

const config = IS_PROD
  ? {
      ...DEFAULT_CONFIG,
      ...PRODUCTION_CONFIG,
    }
  : {
      ...DEFAULT_CONFIG,
    };

var r = require('rethinkdbdash')(config);

module.exports = { db: r };

// /**
//  * Database setup is done here
//  */
// const fs = require('fs');
// const path = require('path');
// const IS_PROD = !process.env.FORCE_DEV && process.env.NODE_ENV === 'production';
//
// const DEFAULT_CONFIG = {
//   db: 'spectrum',
// };
//
// // COMPOSE_RETHINKDB_URL="aws-us-east-1-portal.6.dblayer.com"
// // COMPOSE_RETHINKDB_PORT=19241
// // COMPOSE_RETHINKDB_PASSWORD="12460d0b-b1dc-4505-9cd5-7a96e58ad825"
//
// const PRODUCTION_CONFIG = {
//   password: "12460d0b-b1dc-4505-9cd5-7a96e58ad825",
//   host: "aws-us-east-1-portal.6.dblayer.com"  ,
//   port: 19241,
//   ssl: {
//     ca: !IS_PROD && require('raw-loader!../../cacert'),
//   },
// };
//
// const config = !IS_PROD
//   ? {
//       ...DEFAULT_CONFIG,
//       ...PRODUCTION_CONFIG,
//     }
//   : {
//       ...DEFAULT_CONFIG,
//     };
//
// var r = require('rethinkdbdash')(config);
//
// module.exports = { db: r };
