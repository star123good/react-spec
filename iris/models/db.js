/**
 * Database setup is done here
 */
const fs = require('fs');
const path = require('path');
const IS_PROD = !process.env.FORCE_DEV && process.env.NODE_ENV === 'production';

const config = {
  db: 'spectrum',
  password: 'V3GyuVa4JdKFcMwinfff4FAhRubFJKLNiZFcwg4s6NuRxao2hyhcB2Vbrb4Y',
  host: 'ec2-54-161-144-68.compute-1.amazonaws.com',
  // port: process.env.COMPOSE_RETHINKDB_PORT,
};

var r = require('rethinkdbdash')(config);

module.exports = { db: r };
