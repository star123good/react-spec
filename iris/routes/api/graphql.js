// @flow
import { graphqlExpress } from 'graphql-server-express';
import OpticsAgent from 'optics-agent';
import createLoaders from '../../loaders/';

import createErrorFormatter from '../../utils/create-graphql-error-formatter';
import schema from '../../schema';

export default graphqlExpress(req => ({
  schema,
  formatError: createErrorFormatter(req),
  context: {
    user: req.user,
    loaders: createLoaders(),
    opticsContext: OpticsAgent.context(req),
  },
}));
