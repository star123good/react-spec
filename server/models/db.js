/**
 * Database setup is done here
 */
var r = require('rethinkdbdash')({
  db: 'spectrum',
});

module.exports = { db: r };
