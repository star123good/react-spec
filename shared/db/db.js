// @flow
/**
 * Database setup is done here
 */
import fs from 'fs';
import path from 'path';

const IS_PROD = !process.env.FORCE_DEV && process.env.NODE_ENV === 'production';

const DEFAULT_CONFIG = {
  // Connect to the test database when, well, testing
  db: !process.env.TEST_DB ? 'spectrum' : 'testing',
  max:
    process.env.SENTRY_NAME === 'api' || process.env.SENTRY_NAME === 'hyperion'
      ? 60
      : 1, // Maximum number of connections, default is 1000
  buffer: 1, // Minimum number of connections open at any given moment, default is 50
  timeoutGb: 60 * 1000, // How long should an unused connection stick around, default is an hour, this is a minute
};

let ca;

try {
  ca = fs.readFileSync(path.join(process.cwd(), 'cacert'));
} catch (err) {
  if (IS_PROD) throw err;
}

const PRODUCTION_CONFIG = {
  password: process.env.COMPOSE_RETHINKDB_PASSWORD,
  host: process.env.COMPOSE_RETHINKDB_URL,
  port: process.env.COMPOSE_RETHINKDB_PORT,
  ...(ca
    ? {
        ssl: {
          ca,
        },
      }
    : {}),
};

const config = IS_PROD
  ? {
      ...DEFAULT_CONFIG,
      ...PRODUCTION_CONFIG,
    }
  : {
      ...DEFAULT_CONFIG,
    };

var r = require('rethinkdbdash')(config);

// Exit the process on unhealthy db in test env
if (process.env.TEST_DB) {
  r.getPoolMaster().on('healthy', healthy => {
    if (!healthy) {
      process.exit(1);
    }
  });
}

if (process.env.NODE_ENV === 'development' && process.env.TRACK_DB_PERF) {
  const fs = require('fs');
  const inspect = require('rethinkdb-inspector');
  const queries = [];
  inspect(r, {
    onQueryComplete: (query, { size, time }) => {
      if (query.indexOf('.changes') > -1) return;
      queries.push({ query, time, size });
      fs.writeFileSync(
        'queries-by-time.js',
        JSON.stringify(queries.sort((a, b) => b.time - a.time), null, 2)
      );
      fs.writeFileSync(
        'queries-by-response-size.js',
        JSON.stringify(queries.sort((a, b) => b.size - a.size), null, 2)
      );
    },
  });
}

module.exports = { db: r };
