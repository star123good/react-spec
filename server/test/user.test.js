import { graphql } from 'graphql';
import { db as mockTestDb, setup, teardown, data } from './db';
jest.mock('../models/db', () => ({
  db: mockTestDb,
}));

import schema from '../schema';

describe('queries', () => {
  beforeAll(() => setup(mockTestDb));
  afterAll(() => teardown(mockTestDb));

  it('should fetch a user', () => {
    const query = /* GraphQL */ `
			{
				user(id: "first-user") {
					uid
					createdAt
					lastSeen
					photoURL
					displayName
					username
					email
				}
			}
		`;

    expect.assertions(1);
    return graphql(schema, query).then(result => {
      expect(result).toMatchSnapshot();
    });
  });

  it('should return null for a non-existant id', () => {
    const query = /* GraphQL */ `
			{
				user(id: "non-existant") {
					uid
				}
			}
		`;

    expect.assertions(1);
    return graphql(schema, query).then(result => {
      expect(result).toMatchSnapshot();
    });
  });

  it.skip('fetches a users communities');
  it.skip('fetches a users frequencies');
});
