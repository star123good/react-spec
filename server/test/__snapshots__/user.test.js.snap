exports[`queries should fetch a user 1`] = `
Object {
  "data": Object {
    "user": Object {
      "createdAt": "2017-01-01T23:00:00.000Z",
      "displayName": "First User",
      "email": "first.user@gmail.com",
      "lastSeen": "2017-02-01T23:00:00.000Z",
      "photoURL": "my-photo.jpg",
      "uid": "first-user",
      "username": "first",
    },
  },
}
`;

exports[`queries should return null for a non-existant id 1`] = `
Object {
  "data": Object {
    "user": null,
  },
}
`;
